File: android/src/main/java/com/superwallreactnative/models/PaywallInfo.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.paywall.presentation.PaywallInfo

fun PaywallInfo.toJson(): ReadableMap {
  val map = Arguments.createMap()

  map.putString("identifier", this.identifier)
  map.putString("name", this.name)
  map.putString("url", this.url.toString())
  this.experiment?.let {
    map.putMap("experiment", Experiment.toJson(it))
  }

  val productsArray = Arguments.createArray()
  this.products.forEach { product ->
    val productMap = Arguments.createMap()
    productMap.putString("type", product.type.toString())
    productMap.putString("id", product.fullProductId)
    productsArray.pushMap(productMap)
  }
  map.putArray("products", productsArray)

  val productIdsArray = Arguments.createArray()
  this.productIds.forEach { productId ->
    productIdsArray.pushString(productId)
  }
  map.putArray("productIds", productIdsArray)

  this.presentedByEventWithName?.let { map.putString("presentedByEventWithName", it) }
  this.presentedByEventWithId?.let { map.putString("presentedByEventWithId", it) }
  this.presentedByEventAt?.let { map.putString("presentedByEventAt", it) }
  map.putString("presentedBy", this.presentedBy)
  this.presentationSourceType?.let { map.putString("presentationSourceType", it) }
  this.responseLoadStartTime?.let { map.putString("responseLoadStartTime", it) }
  this.responseLoadCompleteTime?.let { map.putString("responseLoadCompleteTime", it) }
  this.responseLoadFailTime?.let { map.putString("responseLoadFailTime", it) }
  this.responseLoadDuration?.let { map.putDouble("responseLoadDuration", it.toDouble()) }

  map.putBoolean("isFreeTrialAvailable", this.isFreeTrialAvailable)
  map.putString("featureGatingBehavior", this.featureGatingBehavior.rawValue())
  map.putString("closeReason", this.closeReason.rawValue())

  this.webViewLoadStartTime?.let { map.putString("webViewLoadStartTime", it) }
  this.webViewLoadCompleteTime?.let { map.putString("webViewLoadCompleteTime", it) }
  this.webViewLoadFailTime?.let { map.putString("webViewLoadFailTime", it) }
  this.webViewLoadDuration?.let { map.putDouble("webViewLoadDuration", it.toDouble()) }

  this.productsLoadStartTime?.let { map.putString("productsLoadStartTime", it) }
  this.productsLoadCompleteTime?.let { map.putString("productsLoadCompleteTime", it) }
  this.productsLoadFailTime?.let { map.putString("productsLoadFailTime", it) }
  this.productsLoadDuration?.let { map.putDouble("productsLoadDuration", it.toDouble()) }

  this.paywalljsVersion?.let { map.putString("paywalljsVersion", it) }

  val computedPropertyRequestsArray = Arguments.createArray()
  this.computedPropertyRequests.forEach { request ->
    val computedPropertyRequestsMap = Arguments.createMap()
    computedPropertyRequestsMap.putString("placementName", request.eventName)
    computedPropertyRequestsMap.putString("type", request.type.toString())
    computedPropertyRequestsArray.pushMap(computedPropertyRequestsMap)
  }
  map.putArray("computedPropertyRequests", computedPropertyRequestsArray)

  val surveysArray = Arguments.createArray()
  this.surveys.forEach { survey ->
    val surveysMap = Arguments.createMap()
    surveysMap.putString("id", survey.id)
    surveysMap.putString("message", survey.message)
    surveysMap.putString("title", survey.title)
    surveysMap.putString("assignmentKey", survey.assignmentKey)
    surveysMap.putBoolean("includeCloseOption", survey.includeCloseOption)
    surveysMap.putBoolean("includeOtherOption", survey.includeOtherOption)
    surveysMap.putDouble("presentationProbability", survey.presentationProbability)
    surveysMap.putString("presentationCondition", survey.presentationCondition.rawValue)

    val optionsArray = Arguments.createArray()
    survey.options.forEach { option ->
      val optionsMap = Arguments.createMap()
      optionsMap.putString("id", option.id)
      optionsMap.putString("title", option.title)
      optionsArray.pushMap(optionsMap)
    }
    surveysMap.putArray("options", optionsArray)

    surveysArray.pushMap(surveysMap)
  }
  map.putArray("surveys", surveysArray)

  val localNotificationsArray = Arguments.createArray()
  this.localNotifications.forEach { notification ->
    val notificationsMap = Arguments.createMap()
    notificationsMap.putInt("id", notification.id)
    notificationsMap.putString("title", notification.title)
    notificationsMap.putString("body", notification.body)
    notificationsMap.putString("type", notification.type.toJson())
    notificationsMap.putDouble("delay", notification.delay.toDouble())
    localNotificationsArray.pushMap(notificationsMap)
  }
  map.putArray("localNotifications", localNotificationsArray)

  return map
}

File: src/public/PaywallInfo.ts
import { ComputedPropertyRequest } from './ComputedPropertyRequest';
import { Experiment } from './Experiment';
import {
  FeatureGatingBehavior,
  featureGatingBehaviorFromJson,
} from './FeatureGatingBehavior';
import { LocalNotification } from './LocalNotification';
import { PaywallCloseReason } from './PaywallCloseReason';
import { Product } from './Product';
import { Survey } from './Survey';

export class PaywallInfo {
  identifier: string;
  name: string;
  url: string;
  experiment?: Experiment;
  products: Product[];
  productIds: string[];
  presentedByEventWithName?: string;
  presentedByEventWithId?: string;
  presentedByEventAt?: string;
  presentedBy: string;
  presentationSourceType?: string;
  responseLoadStartTime?: string;
  responseLoadCompleteTime?: string;
  responseLoadFailTime?: string;
  responseLoadDuration?: number;
  webViewLoadStartTime?: string;
  webViewLoadCompleteTime?: string;
  webViewLoadFailTime?: string;
  webViewLoadDuration?: number;
  productsLoadStartTime?: string;
  productsLoadCompleteTime?: string;
  productsLoadFailTime?: string;
  productsLoadDuration?: number;
  paywalljsVersion?: string;
  isFreeTrialAvailable: boolean;
  featureGatingBehavior: FeatureGatingBehavior; // Assuming FeatureGatingBehavior is an enum defined elsewhere
  closeReason: PaywallCloseReason; // Assuming PaywallCloseReason is an enum defined elsewhere
  localNotifications: LocalNotification[]; // Assuming LocalNotification is defined elsewhere
  computedPropertyRequests: ComputedPropertyRequest[]; // Assuming ComputedPropertyRequest is defined elsewhere
  surveys: Survey[]; // Assuming Survey is defined elsewhere

  constructor({
    identifier,
    name,
    url,
    experiment,
    products,
    productIds,
    presentedByEventWithName,
    presentedByEventWithId,
    presentedByEventAt,
    presentedBy,
    presentationSourceType,
    responseLoadStartTime,
    responseLoadCompleteTime,
    responseLoadFailTime,
    responseLoadDuration,
    webViewLoadStartTime,
    webViewLoadCompleteTime,
    webViewLoadFailTime,
    webViewLoadDuration,
    productsLoadStartTime,
    productsLoadCompleteTime,
    productsLoadFailTime,
    productsLoadDuration,
    paywalljsVersion,
    isFreeTrialAvailable,
    featureGatingBehavior,
    closeReason,
    localNotifications,
    computedPropertyRequests,
    surveys,
  }: {
    identifier: string,
    name: string,
    url: string,
    experiment?: Experiment,
    products: Product[],
    productIds: string[],
    presentedByEventWithName?: string,
    presentedByEventWithId?: string,
    presentedByEventAt?: string,
    presentedBy: string,
    presentationSourceType?: string,
    responseLoadStartTime?: string,
    responseLoadCompleteTime?: string,
    responseLoadFailTime?: string,
    responseLoadDuration?: number,
    webViewLoadStartTime?: string,
    webViewLoadCompleteTime?: string,
    webViewLoadFailTime?: string,
    webViewLoadDuration?: number,
    productsLoadStartTime?: string,
    productsLoadCompleteTime?: string,
    productsLoadFailTime?: string,
    productsLoadDuration?: number,
    paywalljsVersion?: string,
    isFreeTrialAvailable: boolean,
    featureGatingBehavior: FeatureGatingBehavior,
    closeReason: PaywallCloseReason,
    localNotifications: LocalNotification[],
    computedPropertyRequests: ComputedPropertyRequest[],
    surveys: Survey[],
  }) {
    this.identifier = identifier;
    this.name = name;
    this.url = url;
    this.experiment = experiment;
    this.products = products;
    this.productIds = productIds;
    this.presentedByEventWithName = presentedByEventWithName;
    this.presentedByEventWithId = presentedByEventWithId;
    this.presentedByEventAt = presentedByEventAt;
    this.presentedBy = presentedBy;
    this.presentationSourceType = presentationSourceType;
    this.responseLoadStartTime = responseLoadStartTime;
    this.responseLoadCompleteTime = responseLoadCompleteTime;
    this.responseLoadFailTime = responseLoadFailTime;
    this.responseLoadDuration = responseLoadDuration;
    this.webViewLoadStartTime = webViewLoadStartTime;
    this.webViewLoadCompleteTime = webViewLoadCompleteTime;
    this.webViewLoadFailTime = webViewLoadFailTime;
    this.webViewLoadDuration = webViewLoadDuration;
    this.productsLoadStartTime = productsLoadStartTime;
    this.productsLoadCompleteTime = productsLoadCompleteTime;
    this.productsLoadFailTime = productsLoadFailTime;
    this.productsLoadDuration = productsLoadDuration;
    this.paywalljsVersion = paywalljsVersion;
    this.isFreeTrialAvailable = isFreeTrialAvailable;
    this.featureGatingBehavior = featureGatingBehavior;
    this.closeReason = closeReason;
    this.localNotifications = localNotifications;
    this.computedPropertyRequests = computedPropertyRequests;
    this.surveys = surveys;
  }

  static fromJson(json: any): PaywallInfo {
    return new PaywallInfo({
      identifier: json.identifier,
      name: json.name,
      url: json.url,
      experiment: json.experiment
        ? Experiment.fromJson(json.experiment)
        : undefined,
      products: json.products.map((p: any) => Product.fromJson(p)),
      productIds: json.productIds,
      presentedByEventWithName: json.presentedByEventWithName,
      presentedByEventWithId: json.presentedByEventWithId,
      presentedByEventAt: json.presentedByEventAt,
      presentedBy: json.presentedBy,
      presentationSourceType: json.presentationSourceType,
      responseLoadStartTime: json.responseLoadStartTime,
      responseLoadCompleteTime: json.responseLoadCompleteTime,
      responseLoadFailTime: json.responseLoadFailTime,
      responseLoadDuration: json.responseLoadDuration,
      webViewLoadStartTime: json.webViewLoadStartTime,
      webViewLoadCompleteTime: json.webViewLoadCompleteTime,
      webViewLoadFailTime: json.webViewLoadFailTime,
      webViewLoadDuration: json.webViewLoadDuration,
      productsLoadStartTime: json.productsLoadStartTime,
      productsLoadCompleteTime: json.productsLoadCompleteTime,
      productsLoadFailTime: json.productsLoadFailTime,
      productsLoadDuration: json.productsLoadDuration,
      paywalljsVersion: json.paywalljsVersion,
      isFreeTrialAvailable: json.isFreeTrialAvailable,
      featureGatingBehavior: featureGatingBehaviorFromJson(
        json.featureGatingBehavior
      ),
      closeReason: PaywallCloseReason.fromJson(json.closeReason),
      localNotifications: json.localNotifications.map((n: any) =>
        LocalNotification.fromJson(n)
      ),
      computedPropertyRequests: json.computedPropertyRequests.map((r: any) =>
        ComputedPropertyRequest.fromJson(r)
      ),
      surveys: json.surveys.map((s: any) => Survey.fromJson(s)),
    });
  }
}
